akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -142,6 +142,62 @@ private[akka] final class HeadOptionSink[In](val attributes: Attributes, shape:
 
 /**
  * INTERNAL API
+ */
+private[akka] object LastSink {
+  final class LastOptionSinkSubscriber[In] extends Subscriber[In] {
+    private[this] var prev: In = null.asInstanceOf[In]
+    private[this] var subscription: Subscription = null
+    private[this] val promise: Promise[Option[In]] = Promise[Option[In]]()
+    def future: Future[Option[In]] = promise.future
+    override def onSubscribe(s: Subscription): Unit = {
+      ReactiveStreamsCompliance.requireNonNullSubscription(s)
+      if (subscription ne null) s.cancel()
+      else {
+        subscription = s
+        s.request(Long.MaxValue)
+      }
+    }
+
+    override def onNext(elem: In): Unit = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
+      prev = elem
+    }
+
+    override def onError(t: Throwable): Unit = {
+      ReactiveStreamsCompliance.requireNonNullException(t)
+      promise.tryFailure(t)
+      subscription = null
+    }
+
+    override def onComplete(): Unit = {
+      promise.trySuccess(Option(prev))
+      prev = null.asInstanceOf[In]
+      subscription = null
+    }
+  }
```

Comment:
That's a good comment. I think in general we want to cut down on the RS
implementations, instead opting for reuse. I can take a stab at having it
TCKed.

-- 
Cheers,
√
On 18 Nov 2015 13:34, "Konrad Malawski" <notifications@github.com> wrote:

> In akka-stream/src/main/scala/akka/stream/impl/Sinks.scala
> <https://github.com/akka/akka/pull/18951#discussion_r45193316>:
>
> > +      ReactiveStreamsCompliance.requireNonNullElement(elem)
> > +      prev = elem
> > +    }
> > +
> > +    override def onError(t: Throwable): Unit = {
> > +      ReactiveStreamsCompliance.requireNonNullException(t)
> > +      promise.tryFailure(t)
> > +      subscription = null
> > +    }
> > +
> > +    override def onComplete(): Unit = {
> > +      promise.trySuccess(Option(prev))
> > +      prev = null.asInstanceOf[In]
> > +      subscription = null
> > +    }
> > +  }
>
> raw Subscriber implementation would be good to TCK test (we do still bump
> into silly mistakes with them ince in a while) though not sure how many
> tests such one would pass (semantics wise), have you had a look?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/18951/files#r45193316>.
>